src/nodeViews/createReactNodeViewConstructor.tsx

```diff
@@ -207,11 +273,31 @@ export function createReactNodeViewConstructor(
       />
     );
 
-    const unregisterElement = registerElement(
-      element,
-      dom as HTMLElement,
-      Math.floor(Math.random() * 0xffffff).toString(16)
-    );
+    if (contentDOM) {
+      // Store this node view's registration key on the contentDOM
+      // so it can be retrieved by its children.
+      contentDOM.dataset[REACT_PM_KEY] = key;
+    }
+
+    let unregisterElement: UnregisterElement;
+
+    // ProseMirror hasn't assigned finished constructing the
+    // node view descriptor tree yet, so attempts to ascend it
+    // will fail until after the current call stack has finished
+    // executing.
+    //
+    // Push registration onto the microtask queue so that it will
+    // be executed at the end of the current event loop, after
+    // ProseMirror has finished constructing the node view descriptor
+    // tree.
+    ensureMicrotask(() => {
+      unregisterElement = registerElement(
+        findNearestRegistryKey(editorView, getPos()),
+        element,
+        dom as HTMLElement,
+        key
+      );
+    });
```

Comment:
This is great. Do we have to worry about the nodeview getting destroyed before this runs? I'm not sure it's possible, given that we're in a batch context right now, but it honestly hurts my brain and I'm sorry.

Reply:
Hahahaha I was wondering the same... It doesn't seem possible, right? Like, presumably by the time this runs, all of the state updates have "settled" (because batch context), so any destruction couldn't be triggered until the _next_ event loop. At least, that's what it feels like. It should be pretty obvious if this does break in that way (if "unregisterElement is not callable", this is why), though, so maybe it's ok to assume for now and account for it if it's possible to mess up? Just anecdotally, I didn't have any issues testing it with the demo app.

Reply:
I'm not sure I would know what to do if it destroy was called before the microtask runs. I guess we would maybe want to "cancel" the microtask, which could look like removing the task from the taskQueue? Maybe that's not so bad, actually

Reply:
Yeah, exactly.